### DISCUSSION TITLE
client upgrade to 5.9.8 doesn't fail on a cert mismatch any more and also crashes

### DISCUSSION THREAD

**Discussion by bschue99** [strongswan/strongswan#1541] [Q&A]
I have a problem when upgrading the client to strongswan 5.9.8 / 5.9.9 and requesting a server certificate that the server doesn't provide.
(using EC-CA certificates on the server and client, but requesting a RSA CA certificate from server)

with 5.9.7 and prior, we get the expected connection drop: (sending fatal TLS alert 'certificate unknown'),
newer versions just continue and crash..

any thoughts ?

00[DMN] Starting IKE charon daemon (strongSwan 5.9.7, Linux 5.10.142, aarch64)
00[CFG]   loaded EAP secret for theobroma
00[LIB] loading feature CUSTOM:vici in plugin 'vici'
00[LIB] loading feature CUSTOM:updown in plugin 'updown'
00[LIB] loading feature EAP_SERVER:ID in plugin 'eap-identity'
00[LIB] loading feature EAP_CLIENT:ID in plugin 'eap-identity'
00[LIB] loading feature EAP_SERVER:MSCHAPV2 in plugin 'eap-mschapv2'
00[LIB] loading feature EAP_CLIENT:MSCHAPV2 in plugin 'eap-mschapv2'
00[LIB] loading feature EAP_SERVER:TLS in plugin 'eap-tls'
00[LIB] loading feature EAP_CLIENT:TLS in plugin 'eap-tls'
00[LIB] loading feature XAUTH_SERVER:generic in plugin 'xauth-generic'
00[LIB] loading feature XAUTH_CLIENT:generic in plugin 'xauth-generic'
00[LIB] loading feature CUSTOM:systime-fix in plugin 'systime-fix'
00[CFG] enabling systime-fix, threshold: Tue Jan  1 00:00:00 2999
00[CFG]
00[LIB] loading feature CUSTOM:led in plugin 'led'
00[LIB] loaded plugins: charon gcm pkcs11 aes des rc2 sha2 sha1 md4 md5 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8 fips-prf curve25519 xcbc cmac hmac kdf drbg curl attr kernel-netlink resolve socket-default bypass-lan stroke vici updown eap-identity eap-mschapv2 eap-tls xauth-generic systime-fix led counters
00[LIB] unable to load 5 plugin features (5 due to unmet dependencies)
00[LIB] dropped capabilities, running as uid 102, gid 105
00[JOB] spawning 16 worker threads
09[CFG] vici client 1 requests: load-cert
09[CFG] loaded certificate 'CN=DEVICE-IPSEC-SECP384R1-PEER-DEBUG, CN=Munich, C=DE'
16[CFG] vici client 1 requests: load-authority
16[CFG]   cacert = C=DE, CN=Munich, CN=DEVICE-IPSEC-RSA4K-CA-DEBUG
16[CFG] loaded ANY private key
16[CFG] vici client 1 requests: initiate
16[CFG] vici initiate CHILD_SA 'tunnel'
05[IKE] initiating IKE_SA connect_with_eap[1] to 10.2.0.1
05[IKE] IKE_SA connect_with_eap[1] state change: CREATED => CONNECTING
05[CFG] configured proposals: IKE:AES_GCM_8_256/PRF_HMAC_SHA2_256/PRF_HMAC_SHA2_384/PRF_HMAC_SHA2_512/ECP_384/ECP_521/ECP_384_BP/ECP_512_BP/MODP_4096, IKE:AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/PRF_HMAC_SHA2_256/PRF_HMAC_SHA2_384/PRF_HMAC_SHA2_512/ECP_384/ECP_521/ECP_384_BP/ECP_512_BP/MODP_4096
05[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
05[NET] sending packet: from 10.3.99.99[500] to 10.2.0.1[500] (452 bytes)
07[NET] received packet: from 10.2.0.1[500] to 10.3.99.99[500] (357 bytes)
07[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
07[IKE] received FRAGMENTATION_SUPPORTED notify
07[IKE] received SIGNATURE_HASH_ALGORITHMS notify
07[IKE] received CHILDLESS_IKEV2_SUPPORTED notify
07[CFG] selecting proposal:
07[CFG]   no acceptable ENCRYPTION_ALGORITHM found
07[CFG] selecting proposal:
07[CFG]   proposal matches
07[CFG] received proposals: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_384
07[CFG] configured proposals: IKE:AES_GCM_8_256/PRF_HMAC_SHA2_256/PRF_HMAC_SHA2_384/PRF_HMAC_SHA2_512/ECP_384/ECP_521/ECP_384_BP/ECP_512_BP/MODP_4096, IKE:AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/PRF_HMAC_SHA2_256/PRF_HMAC_SHA2_384/PRF_HMAC_SHA2_512/ECP_384/ECP_521/ECP_384_BP/ECP_512_BP/MODP_4096
07[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_384
07[CFG] received supported signature hash algorithms: sha256 sha384 sha512 identity
07[IKE] received cert request for "C=DE, CN=Munich, CN=DEVICE-IPSEC-RSA4K-CA-DEBUG"
07[IKE] received cert request for unknown ca with keyid 17:17:d4:47:...
07[IKE] received 1 cert requests for an unknown ca
07[IKE] reinitiating already active tasks
07[IKE]   IKE_CERT_PRE task
07[IKE]   IKE_AUTH task
07[IKE] sending cert request for "C=DE, CN=Munich, CN=DEVICE-IPSEC-RSA4K-CA-DEBUG"
07[IKE] building INTERNAL_IP4_DNS attribute
07[CFG] proposing traffic selectors for us:
07[CFG]  0.0.0.0..169.253.255.255
07[CFG]  169.255.0.0..254.255.255.255
07[CFG] proposing traffic selectors for other:
07[CFG]  0.0.0.0/0
07[CFG] configured proposals: ESP:AES_GCM_8_256/NO_EXT_SEQ, ESP:AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/NO_EXT_SEQ
07[IKE] establishing CHILD_SA tunnel{1}
07[ENC] generating IKE_AUTH request 1 [ IDi CERTREQ CPRQ(ADDR DNS) SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
07[NET] sending packet: from 10.3.99.99[4500] to 10.2.0.1[4500] (336 bytes)
11[NET] received packet: from 10.2.0.1[4500] to 10.3.99.99[4500] (160 bytes)
11[ENC] parsed IKE_AUTH response 1 [ IDr EAP/REQ/TLS ]
11[IKE] server requested EAP_TLS authentication (id 0xC6)
11[IKE] allow mutual EAP-only authentication
11[IKE] reinitiating already active tasks
11[IKE]   IKE_CERT_PRE task
11[IKE]   IKE_AUTH task
11[ENC] generating IKE_AUTH request 2 [ EAP/RES/TLS ]
11[NET] sending packet: from 10.3.99.99[4500] to 10.2.0.1[4500] (304 bytes)
06[NET] received packet: from 10.2.0.1[4500] to 10.3.99.99[4500] (1024 bytes)
06[ENC] parsed IKE_AUTH response 2 [ EAP/REQ/TLS ]
06[TLS] processing TLS Handshake record (933 bytes)
06[TLS] received TLS ServerHello handshake (54 bytes)
06[TLS] negotiated TLS 1.2 using suite TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
06[TLS] received TLS Certificate handshake (581 bytes)
06[LIB] builder L0 CRED_CERTIFICATE - X509 of plugin 'x509'
06[LIB] builder L1 CRED_PUBLIC_KEY - ANY of plugin 'pkcs1'
06[LIB] builder L2 CRED_PUBLIC_KEY - ECDSA of plugin 'openssl'
06[TLS] received TLS server certificate 'CN=DEVICE-IPSEC-SECP384R1-SITE-DEBUG, CN=Munich, C=DE'
06[TLS] received TLS ServerKeyExchange handshake (177 bytes)
06[CFG]   using certificate "CN=DEVICE-IPSEC-SECP384R1-SITE-DEBUG, CN=Munich, C=DE"
06[CFG]   certificate "CN=DEVICE-IPSEC-SECP384R1-SITE-DEBUG, CN=Munich, C=DE" key: 384 bit ECDSA
06[CFG] no issuer certificate found for "CN=DEVICE-IPSEC-SECP384R1-SITE-DEBUG, CN=Munich, C=DE"
06[CFG]   issuer is "C=DE, CN=Munich, CN=DEVICE-IPSEC-SECP384R1-CA-DEBUG"
06[TLS] no TLS public key found for server 'CN=DEVICE-IPSEC-SECP384R1-SITE-DEBUG, CN=Munich, C=DE'
06[TLS] sending fatal TLS alert 'certificate unknown'     <-------- fails as expected ---------->
06[TLS] sending TLS Alert record (2 bytes)
06[TLS] sending EAP_TLS packet (17 bytes)
06[IKE] reinitiating already active tasks
06[IKE]   IKE_CERT_PRE task
06[IKE]   IKE_AUTH task
06[ENC] generating IKE_AUTH request 3 [ EAP/RES/TLS ]
06[NET] sending packet: from 10.3.99.99[4500] to 10.2.0.1[4500] (96 bytes)
08[NET] received packet: from 10.2.0.1[4500] to 10.3.99.99[4500] (80 bytes)
08[ENC] parsed IKE_AUTH response 3 [ EAP/FAIL ]
08[IKE] received EAP_FAILURE, EAP authentication failed
08[ENC] generating INFORMATIONAL request 4 [ N(AUTH_FAILED) ]
08[NET] sending packet: from 10.3.99.99[4500] to 10.2.0.1[4500] (80 bytes)
08[IKE] IKE_SA connect_with_eap[1] state change: CONNECTING => DESTROYING

bad case:

00[DMN] Starting IKE charon daemon (strongSwan 5.9.8, Linux 5.10.142, aarch64)
00[LIB] loading feature CUSTOM:vici in plugin 'vici'
00[LIB] loading feature CUSTOM:updown in plugin 'updown'
00[LIB] loading feature EAP_SERVER:ID in plugin 'eap-identity'
00[LIB] loading feature EAP_CLIENT:ID in plugin 'eap-identity'
00[LIB] loading feature EAP_SERVER:MSCHAPV2 in plugin 'eap-mschapv2'
00[LIB] loading feature EAP_CLIENT:MSCHAPV2 in plugin 'eap-mschapv2'
00[LIB] loading feature EAP_SERVER:TLS in plugin 'eap-tls'
00[LIB] loading feature EAP_CLIENT:TLS in plugin 'eap-tls'
00[LIB] loading feature XAUTH_SERVER:generic in plugin 'xauth-generic'
00[LIB] loading feature XAUTH_CLIENT:generic in plugin 'xauth-generic'
00[LIB] loading feature CUSTOM:systime-fix in plugin 'systime-fix'
00[CFG] enabling systime-fix, threshold: Tue Jan  1 00:00:00 2999
00[CFG]
00[LIB] loading feature CUSTOM:led in plugin 'led'
00[LIB] loaded plugins: charon pkcs11 aes des rc2 sha2 sha1 md4 md5 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8 fips-prf curve25519 xcbc cmac hmac kdf gcm drbg curl attr kernel-netlink resolve socket-default bypass-lan stroke vici updown eap-identity eap-mschapv2 eap-tls xauth-generic systime-fix led counters
00[LIB] unable to load 5 plugin features (5 due to unmet dependencies)
00[LIB] dropped capabilities, running as uid 102, gid 105
00[JOB] spawning 16 worker threads
01[LIB] created thread 01 [8787]
02[LIB] created thread 02 [8788]
03[LIB] created thread 03 [8789]
04[LIB] created thread 04 [8786]
05[LIB] created thread 05 [8790]
06[LIB] created thread 06 [8791]
07[LIB] created thread 07 [8792]
08[LIB] created thread 08 [8793]
09[LIB] created thread 09 [8794]
10[LIB] created thread 10 [8795]
11[LIB] created thread 11 [8796]
12[LIB] created thread 12 [8797]
13[LIB] created thread 13 [8798]
14[LIB] created thread 14 [8799]
15[LIB] created thread 15 [8800]
16[LIB] created thread 16 [8785]
05[IKE] installed bypass policy for 169.254.0.0/16
06[CFG] vici client 1 connected
08[CFG] vici client 1 requests: load-cert
08[CFG] loaded certificate 'CN=DEVICE-IPSEC-SECP384R1-PEER-DEBUG, CN=Munich, C=DE'
12[CFG] vici client 1 requests: load-authority
12[CFG]  authority vpn-ca-rsa4k:
12[CFG]   cacert = C=DE, CN=Munich, CN=DEVICE-IPSEC-RSA4K-CA-DEBUG
12[CFG] vici client 1 requests: load-conn
12[CFG] added vici connection: mypass_connection
12[CFG] installing 'child_A'
12[CFG] installing 'child_B'
12[CFG] vici client 1 requests: load-key
12[CFG] loaded ANY private key
12[CFG] vici client 1 requests: initiate
12[CFG] vici initiate CHILD_SA 'tunnel'
13[IKE] initiating IKE_SA connect_with_eap[1] to 10.2.0.1
13[IKE] IKE_SA connect_with_eap[1] state change: CREATED => CONNECTING
13[CFG] configured proposals: IKE:AES_GCM_8_256/PRF_HMAC_SHA2_256/PRF_HMAC_SHA2_384/PRF_HMAC_SHA2_512/ECP_384/ECP_521/ECP_384_BP/ECP_512_BP/MODP_4096, IKE:AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/PRF_HMAC_SHA2_256/PRF_HMAC_SHA2_384/PRF_HMAC_SHA2_512/ECP_384/ECP_521/ECP_384_BP/ECP_512_BP/MODP_4096
13[CFG] sending supported signature hash algorithms: sha256 sha384 sha512 identity
13[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
13[NET] sending packet: from 10.3.99.99[500] to 10.2.0.1[500] (452 bytes)
14[NET] received packet: from 10.2.0.1[500] to 10.3.99.99[500] (357 bytes)
14[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
14[IKE] received FRAGMENTATION_SUPPORTED notify
14[IKE] received SIGNATURE_HASH_ALGORITHMS notify
14[IKE] received CHILDLESS_IKEV2_SUPPORTED notify
14[CFG] selecting proposal:
14[CFG]   no acceptable ENCRYPTION_ALGORITHM found
14[CFG] selecting proposal:
14[CFG]   proposal matches
14[CFG] received proposals: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_384
14[CFG] configured proposals: IKE:AES_GCM_8_256/PRF_HMAC_SHA2_256/PRF_HMAC_SHA2_384/PRF_HMAC_SHA2_512/ECP_384/ECP_521/ECP_384_BP/ECP_512_BP/MODP_4096, IKE:AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/PRF_HMAC_SHA2_256/PRF_HMAC_SHA2_384/PRF_HMAC_SHA2_512/ECP_384/ECP_521/ECP_384_BP/ECP_512_BP/MODP_4096
14[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_384
14[CFG] received supported signature hash algorithms: sha256 sha384 sha512 identity
14[IKE] received cert request for "C=DE, CN=Munich, CN=DEVICE-IPSEC-RSA4K-CA-DEBUG"
14[IKE] received cert request for unknown ca with keyid 17:17:d4:47:...
14[IKE] received 1 cert requests for an unknown ca
14[IKE] reinitiating already active tasks
14[IKE]   IKE_CERT_PRE task
14[IKE]   IKE_AUTH task
14[IKE] sending cert request for "C=DE, CN=Munich, CN=DEVICE-IPSEC-RSA4K-CA-DEBUG"
14[IKE] building INTERNAL_IP4_DNS attribute
14[CFG] proposing traffic selectors for us:
14[CFG]  0.0.0.0..169.253.255.255
14[CFG]  169.255.0.0..254.255.255.255
14[CFG] proposing traffic selectors for other:
14[CFG]  0.0.0.0/0
14[CFG] configured proposals: ESP:AES_GCM_8_256/NO_EXT_SEQ, ESP:AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/NO_EXT_SEQ
14[IKE] establishing CHILD_SA tunnel{1}
14[ENC] generating IKE_AUTH request 1 [ IDi CERTREQ CPRQ(ADDR DNS) SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
14[NET] sending packet: from 10.3.99.99[4500] to 10.2.0.1[4500] (336 bytes)
15[NET] received packet: from 10.2.0.1[4500] to 10.3.99.99[4500] (160 bytes)
15[ENC] parsed IKE_AUTH response 1 [ IDr EAP/REQ/TLS ]
15[IKE] server requested EAP_TLS authentication (id 0x7D)
15[IKE] allow mutual EAP-only authentication
15[IKE] reinitiating already active tasks
15[IKE]   IKE_CERT_PRE task
15[IKE]   IKE_AUTH task
15[ENC] generating IKE_AUTH request 2 [ EAP/RES/TLS ]
15[NET] sending packet: from 10.3.99.99[4500] to 10.2.0.1[4500] (304 bytes)
16[NET] received packet: from 10.2.0.1[4500] to 10.3.99.99[4500] (1024 bytes)
16[ENC] parsed IKE_AUTH response 2 [ EAP/REQ/TLS ]
16[TLS] processing TLS Handshake record (933 bytes)
16[TLS] received TLS ServerHello handshake (54 bytes)
16[TLS] negotiated TLS 1.2 using suite TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
16[TLS] received TLS Certificate handshake (581 bytes)
16[LIB] builder L0 CRED_CERTIFICATE - X509 of plugin 'x509'
16[LIB] builder L1 CRED_PUBLIC_KEY - ANY of plugin 'pkcs1'
16[LIB] builder L2 CRED_PUBLIC_KEY - ECDSA of plugin 'openssl'
16[TLS] received TLS server certificate 'CN=DEVICE-IPSEC-SECP384R1-SITE-DEBUG, CN=Munich, C=DE'
16[TLS] received TLS ServerKeyExchange handshake (177 bytes)
16[CFG]   using certificate "CN=DEVICE-IPSEC-SECP384R1-SITE-DEBUG, CN=Munich, C=DE"
16[CFG]   certificate "CN=DEVICE-IPSEC-SECP384R1-SITE-DEBUG, CN=Munich, C=DE" key: 384 bit ECDSA
16[CFG] no issuer certificate found for "CN=DEVICE-IPSEC-SECP384R1-SITE-DEBUG, CN=Munich, C=DE"
16[CFG]   issuer is "C=DE, CN=Munich, CN=DEVICE-IPSEC-SECP384R1-CA-DEBUG"
16[TLS] verified signature with ECDSA_SHA256        <---------- doesn't fail anymore, ignores missing CA -------->
16[TLS] received TLS CertificateRequest handshake (101 bytes)
16[TLS] received TLS cert request for 'C=DE, CN=Munich, CN=DEVICE-IPSEC-RSA4K-CA-DEBUG
16[TLS] received TLS ServerHelloDone handshake (0 bytes)
16[TLS] sending TLS client certificate 'CN=DEVICE-IPSEC-SECP384R1-PEER-DEBUG, CN=Munich, C=DE'
16[TLS] sending TLS Certificate handshake (580 bytes)
16[TLS] sending TLS ClientKeyExchange handshake (66 bytes)
16[TLS] created signature with ECDSA_SHA384
16[TLS] sending TLS CertificateVerify handshake (108 bytes)
16[TLS] sending TLS Handshake record (766 bytes)
16[TLS] sending TLS ChangeCipherSpec record (1 bytes)
16[TLS] sending TLS Finished handshake (12 bytes)
16[TLS] sending TLS Handshake record (64 bytes)
16[TLS] sending EAP_TLS packet (856 bytes)
16[IKE] reinitiating already active tasks
16[IKE]   IKE_CERT_PRE task
16[IKE]   IKE_AUTH task
16[ENC] generating IKE_AUTH request 3 [ EAP/RES/TLS ]
16[NET] sending packet: from 10.3.99.99[4500] to 10.2.0.1[4500] (928 bytes)
05[NET] received packet: from 10.2.0.1[4500] to 10.3.99.99[4500] (160 bytes)
05[ENC] parsed IKE_AUTH response 3 [ EAP/REQ/TLS ]
05[TLS] processing TLS ChangeCipherSpec record (1 bytes)
05[TLS] processing TLS Handshake record (64 bytes)
05[TLS] received TLS Finished handshake (12 bytes)
05[TLS] sending EAP_TLS acknowledgment packet
05[IKE] reinitiating already active tasks
05[IKE]   IKE_CERT_PRE task
05[IKE]   IKE_AUTH task
05[ENC] generating IKE_AUTH request 4 [ EAP/RES/TLS ]
05[NET] sending packet: from 10.3.99.99[4500] to 10.2.0.1[4500] (80 bytes)
06[NET] received packet: from 10.2.0.1[4500] to 10.3.99.99[4500] (80 bytes)
06[ENC] parsed IKE_AUTH response 4 [ EAP/SUCC ]
06[IKE] EAP method EAP_TLS succeeded, MSK established
06[IKE] reinitiating already active tasks
06[IKE]   IKE_CERT_PRE task
06[IKE]   IKE_AUTH task
06[IKE] authentication of '2b:0b:db:0d:...' (myself) with EAP
06[ENC] generating IKE_AUTH request 5 [ AUTH ]
06[NET] sending packet: from 10.3.99.99[4500] to 10.2.0.1[4500] (112 bytes)
07[NET] received packet: from 10.2.0.1[4500] to 10.3.99.99[4500] (272 bytes)
07[ENC] parsed IKE_AUTH response 5 [ AUTH CPRP(ADDR) SA TSi TSr N(AUTH_LFT) N(MOBIKE_SUP) N(ADD_4_ADDR) N(ADD_6_ADDR) ]
07[IKE] authentication of 'CN=DEVICE-IPSEC-SECP384R1-SITE-DEBUG, CN=Munich, C=DE' with EAP successful
07[DMN] thread 7 received 11
07[LIB]  dumping 1 stack frame addresses:
07[LIB]   linux-vdso.so.1 @ 0x7fa1243000 (__kernel_rt_sigreturn+0x0) [0x7fa12437c0]
07[LIB]     ->
 dumping 1 stack frame addresses:
  linux-vdso.so.1 @ 0x7fa1243000 (__kernel_rt_sigreturn+0x0) [0x7fa12437c0]
addr2line: 'linux-vdso.so.1': No such file
    ->
07[DMN] killing ourself, received critical signal

**Comment by tobiasbrunner** [comment]
Yes, that's a bug in 63fd718915b5d246dcc5560382db0c30de309040. The local `public` variable previously only had a value assigned if a trusted public key was found (with the proper refcount). It was reused to determine the public key type, so it was always set even if the certificate was not trusted. Since no additional reference was taken for the untrusted public key, this caused a double-free later.

~~I've pushed a fix for this to the _1541-tls-pubkey_ branch~~ The fix is actually already on master as I wanted to include it in the latest Android version.

**Reply by bschue99** [reply]
Thanks, using 4d3fc90cafc1ee15e90f7af354ae2270fdce994e,  libtls: Fix double-free for untrusted peer certificates 
works as patch on 5.9.8
